### ISSUE TITLE
[Bug]: SQLite3 shipped with DB Browser for SQLite is not compiled with full Unicode support via SQLITE_ENABLE_ICU

### ISSUE THREAD

**Issue by solodon4** [sqlitebrowser/sqlitebrowser#3708]
### What did you do?

Any SQL query involving UPPER or LOWER on strings that are not from Latin alphabets do not work. The same queries work perfectly fine in sqlite3 I build myself with `-DSQLITE_ENABLE_ICU`, which is not enabled by default.

### What did you expect to see?

I would expect SQL queries on strings in other languages work the same way they do on Latin languages.

### What did you see instead?

Strings in Cyrillic alphabets are unaffected by UPPER and LOWER

### DB4S Version

3.12.2

### What OS are you seeing the problem on?

MacOS

### OS version

MacOS Sonoma 14.5 (23F79)

### Prevention against duplicate issues

- [X] I have searched for similar issues

**Comment by justinclift** [comment]
Ahh, you're using the previous 3.12.2 release.

Is it feasible for you to use the newest release ([3.13.0](https://github.com/sqlitebrowser/sqlitebrowser/releases/tag/v3.13.0)), or one of our builds created from our development sources? ie [these](https://github.com/sqlitebrowser/sqlitebrowser/releases/tag/continuous)

**Comment by solodon4** [comment]
I had searched this GitHub repo for `SQLITE_ENABLE_ICU` when I filed this bug and since I didn't find anything relevant, I assumed the problem hasn't been fixed recently.  I have just installed 3.13.0:

```
DB Browser for SQLite Version 3.13.0
Last commit hash when built: 777fe12

Built for arm64-little_endian-lp64, running on arm64
Qt Version 5.15.13
SQLCipher Version 4.6.0 community (based on SQLite 3.45.3).
```

and can confirm that the problem still exists.

**Comment by lucydodo** [comment]
I'll check the compile options this weekend and add them if possible.

**Comment by lucydodo** [comment]
I was going to add that compile option, but our build system, Ubuntu, limits the build with that option,
so I recommend using the [text extension](https://github.com/nalgeon/sqlean/blob/main/docs/text.md).

```sql
WITH Ukrainian(word) AS 
  (VALUES ('Один'),('Два'),('Три'),('Чотири'),('П''ять'))
SELECT text_lower(word), text_upper(word)
FROM Ukrainian;
```
<img width="780" alt="image" src="https://github.com/user-attachments/assets/94afcd05-d26f-4cc1-ab6f-ec278ffbafd2">

**Comment by solodon4** [comment]
Few questions regarding your suggestions:

- How does one load that extension into a running build of SQLite Browser?
- Do I need to build it myself or is there a repository of those?
- How many users of SQLite Browser are actually aware of how to get and load those extensions?
- I am primarily needing this on Mac, are they available on Mac?
- What is the threshold y'all have for something to become part of SQLite Browser itself vs. having it as an extension? e.g. you went with compiling SQLiteBrowser with the `-DSQLITE_ENABLE_MATH_FUNCTIONS` but now you are suggesting not to go with compiling it with `-DSQLITE_ENABLE_ICU`

Thank you!

**Comment by lucydodo** [comment]
Hi, @solodon4 
1. Select 'DB Browser for SQLite' -> 'Settings' from the top menu bar.
Then in the 'Preferences' window that opens, enable 'sqlean.dylib' by checking it in the 'Extensions' tab.
2. Included by built-in in v3.13.0 stable release or the latest nighty build.
3. We don't know because we don't configure telemetry around it.
4. Yes, I'm a Mac user too.
5. I don't understand this question, can you explain a bit more?

I hope you find my answers helpful, and if you have any questions, please let me know. :)

**Comment by solodon4** [comment]
The last question was about where do you draw the line which SQLite features are available directly in the app out of the box and which should be enabled via extensions? For example, functions `power` and `log` are not available in SQLite by default, you had to compile SQLite library with `-DSQLITE_ENABLE_MATH_FUNCTIONS` for them to become available, which you did. Now, an entirely different subset of functionality related to Unicode is also available in SQLite when you compile it with `-DSQLITE_ENABLE_ICU`, yet you suggest not to build it into the SQLiteBrowser by default, but load as extension. My problem is that the above macro enables a whole lot more in support of Unicode than just support for `lower` and `upper` and for all of that functionality I will have to find (potentially different) extensions, learn which functions I have to use in place of the standard ones etc. It would be much easier and straightforward if you've simply built SQLite library you ship with the app with `-DSQLITE_ENABLE_ICU`.

**Comment by solodon4** [comment]
P.S. If that is of any help, here is how I build SQLite on Linux and Mac to have ICU support:

on Linux:

```shell
# See: https://www.sqlite.org/howtocompile.html and http://userguide.icu-project.org/howtouseicu for details
sudo apt-get install sqlite3 libsqlite3-dev libicu-dev
gcc -O3 -o sqlite3 -DSQLITE_ENABLE_ICU -DSQLITE_ENABLE_MATH_FUNCTIONS sqlite3.c shell.c `pkg-config --libs --cflags icu-uc icu-i18n`
```

on Mac:

```shell
brew install icu4c
brew info icu4c
# copy provided CPPFLAGS and LDFLAGS from the above into below command line:
clang -DSQLITE_ENABLE_ICU -DSQLITE_ENABLE_MATH_FUNCTIONS -I/opt/homebrew/opt/icu4c/include sqlite3.c shell.c -o sqlite3 -L/opt/homebrew/opt/icu4c/lib -licuuc -licui18n
```

**Comment by lucydodo** [comment]
That makes sense. It's a little different from the build instruction I was looking into, though. I read somewhere that on our build system, Ubuntu 24.04, the `libicu-dev` package fails to build because it's missing a few things that build that compilation option. I'll give it a try and add the option if it succeeds, thanks for the valuable information.

**Comment by justinclift** [comment]
> you had to compile SQLite library with -DSQLITE_ENABLE_MATH_FUNCTIONS for them to become available, which you did.

Nope, that was me.  The approach we've taken so far is "if it's easy to add and seems useful for people, lets add it".  It's a pretty loose definition, but seems to work ok when just adding an extra option to the compile line is all that's needed.

For situations where a bunch of time and effort are needed to make it work, and it turns out to have some downsides (of whatever sort) then it needs to be thought about a lot more carefully.

I'm not sure how easily added the ICU support pieces are, but hopefully it does turn out to be pretty simple.  From rough memory of ICU support in other things, my suspicion is that it'll increase the size of the compiled binaries significantly.  Not sure if that's really a problem these days though, with multi-megabyte sized binaries being pretty common (Go, Rust?).

**Comment by bairog** [comment]
@lucydodo Any progress/news? I'm also interested in that feature. But I need it to make LIKE statement work in case-insensitive manner with non-ASCII characters. Something like `SELECT * FROM Ukrainian WHERE word LIKE '%РИ'`: with enabled ICU it will return `'Три'` and `'Чотири'` but now with disabled ICU it returns nothing. As I understand [text extension](https://github.com/nalgeon/sqlean/blob/main/docs/text.md) that you [recommended above](https://github.com/sqlitebrowser/sqlitebrowser/issues/3708#issuecomment-2311906525) cannot help here without unnecessary complication of the query.
BTW. Instead of [text extension](https://github.com/nalgeon/sqlean/blob/main/docs/text.md) you can load [unicode extension](https://github.com/nalgeon/sqlean/blob/main/docs/unicode.md) (also from sqlean extension pack). Is is deprecated but still work. And with it you can use "standard" SQL query - `SELECT * FROM Ukrainian WHERE word LIKE '%РИ'`

**Comment by lucydodo** [comment]
I'll have to take a look at the SQLite build script side of things soon, and I'll try to resolve this issue then.

**Comment by alexey-leonovich** [comment]
@lucydodo Any progress/news?

**Comment by lucydodo** [comment]
@bairog @alexey-leonovich I'm sorry for the delay in this work. I will start on the related task sometime next week.
